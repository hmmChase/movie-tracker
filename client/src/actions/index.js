export const addMovies = movies => ({
  type: 'ADD_MOVIES',
  movies
});

export const toggleLogin = () => ({
  type: 'TOGGLE_LOGIN'
});

export const storeUserData = ({ id, name }) => ({
  type: 'STORE_USER_DATA',
  id,
  name
});

export const loadFavorites = favorites => ({
  type: 'LOAD_FAVORITES',
  favorites
});