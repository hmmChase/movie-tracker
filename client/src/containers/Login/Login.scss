.Login {
  button {
    margin: 2rem auto;
  }
}
