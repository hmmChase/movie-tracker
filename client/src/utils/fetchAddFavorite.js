export const fetchAddFavorite = async (movie, user_id) => {
  try {
    const url = '/api/users/favorites/new';
    const options = {
      method: 'POST',
      body: JSON.stringify({
        ...movie,
        user_id
      }),
      headers: { 'Content-Type': 'application/json' }
    };
    const response = await fetch(url, options);

    if (!response.ok) {
      throw new Error(`${response.status}`);
    }
    return response;
  } catch (error) {
    throw new Error(`Network request failed. (error: ${error.message})`);
  }
};
