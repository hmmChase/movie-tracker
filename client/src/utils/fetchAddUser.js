export const fetchAddUser = async user => {
  try {
    const url = '/api/users/new';
    const options = {
      method: 'POST',
      body: JSON.stringify(user),
      headers: { 'Content-Type': 'application/json' }
    };
    const response = await fetch(url, options);
    if (!response.ok) {
      throw new Error(`${response.status}`);
    }
    return response;
  } catch (error) {
    throw new Error(`Network request failed. (error: ${error.message})`);
  }
};
