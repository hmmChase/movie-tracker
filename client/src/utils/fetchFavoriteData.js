import { doFetch } from './doFetch';

export const fetchFavoriteData = async user_id => {
  const url = `/api/users/${user_id}/favorites`;
  const favoritesList = await doFetch(url);
  return favoritesList.data;
};
