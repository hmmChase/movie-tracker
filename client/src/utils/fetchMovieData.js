import { apiKey } from '../private/apiKey';
import { doFetch } from './doFetch';

export const fetchMovieData = async () => {
  const url = 'https://api.themoviedb.org/3/movie/now_playing?api_key=';
  const dirtyMovieData = await doFetch(url + apiKey);
  return cleanMovieData(dirtyMovieData.results);
};

export const cleanMovieData = movieData => {
  return movieData.map(movie => ({
    movie_id: movie.id,
    title: movie.title,
    poster_path: movie.poster_path,
    release_date: movie.release_date,
    vote_average: movie.vote_average,
    overview: movie.overview
  }));
};
