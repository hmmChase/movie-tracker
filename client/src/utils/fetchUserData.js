import { doFetch } from './doFetch';

export const fetchUserData = async () => await doFetch('/api/users');
