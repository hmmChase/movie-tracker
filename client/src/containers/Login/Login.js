import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { fetchUserData } from '../../utils/fetchUserData';
import { fetchFavoriteData } from '../../utils/fetchFavoriteData';
import { toggleLogin, storeUserData, loadFavorites } from '../../actions';
import './Login.css';

export class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      password: ''
    };
  }

  handleChange = event => {
    const { name, value } = event.target;

    this.setState({
      [name]: value
    });
  };

  handleSubmit = async event => {
    event.preventDefault();
    const userData = await fetchUserData();
    const foundUser = userData.data.find(
      user => user.email === this.state.email.toLowerCase()
    );

    if (this.state.email === "") {
      alert('Please enter an email address');
    } else if (!foundUser) {
      alert('No such user');
    } else if (foundUser.password === this.state.password) {
      const favoritesArray = await fetchFavoriteData(foundUser.id);
      this.props.toggleLogin();
      this.props.storeUserData(foundUser);
      this.props.loadFavorites(favoritesArray);
      this.props.history.push('/');
    } else {
      alert('Incorrect Password');
    }
  };

  render() {
    return (
      <div className="Login">
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            placeholder="Enter Email"
            name="email"
            value={this.state.email}
            onChange={this.handleChange}
            className="inputField"
          />
          <input
            type="password"
            placeholder="Enter Password"
            name="password"
            value={this.state.password}
            onChange={this.handleChange}
            className="inputField"
          />
          <button className="loginButton" />
        </form>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  loggedIn: state.user.loggedIn,
  favorites: state.user.favorites
});

export const mapDispatchToProps = dispatch => ({
  toggleLogin: () => dispatch(toggleLogin()),
  storeUserData: userData => dispatch(storeUserData(userData)),
  loadFavorites: favoritesArray => dispatch(loadFavorites(favoritesArray))
});

Login.propTypes = {
  loggedIn: PropTypes.bool.isRequired,
  favorites: PropTypes.array.isRequired,
  toggleLogin: PropTypes.func.isRequired,
  storeUserData: PropTypes.func.isRequired,
  loadFavorites: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Login));
