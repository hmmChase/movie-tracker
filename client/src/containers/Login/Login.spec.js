import React from 'react';
import { shallow } from 'enzyme';
import { Login, mapStateToProps, mapDispatchToProps } from './Login';
import { toggleLogin, storeUserData, loadFavorites } from '../../actions/';
import { fetchUserData } from '../../utils/fetchUserData';
jest.mock('../../utils/fetchUserData');
import { fetchFavoriteData } from '../../utils/fetchFavoriteData';
jest.mock('../../utils/fetchFavoriteData');

describe('Login', () => {
  let login;
  let mockProps;
  beforeEach(() => {
    mockProps = {
      loggedIn: false,
      favorites: [],
      toggleLogin: jest.fn(),
      storeUserData: jest.fn(),
      loadFavorites: jest.fn(),
      history: { push: jest.fn() }
    };
    login = shallow(<Login {...mockProps} />);
  });

  it('matches snapshot', () => {
    expect(login).toMatchSnapshot();
  });

  describe('handleChange', () => {
    it('sets state.email on change of the email input', () => {
      const mockEvent = {
        target: {
          name: 'email',
          value: 'mock@email.com'
        }
      };
      login.instance().handleChange(mockEvent);

      expect(login.state()).toEqual({ email: 'mock@email.com', password: '' });
    });

    it('sets state.password on change of the password input', () => {
      const mockEvent = {
        target: {
          name: 'password',
          value: 'password'
        }
      };
      login.instance().handleChange(mockEvent);

      expect(login.state()).toEqual({ email: '', password: 'password' });
    });
  });

  describe('handleSubmit', () => {
    it('call fetchUserData', async () => {
      const mockEvent = { preventDefault: jest.fn() };
      const userData = {
        data: [
          {
            id: 1,
            name: 'Taylor',
            password: 'password',
            email: 'tman2272@aol.com'
          }
        ]
      };
      fetchUserData.mockImplementation(() => Promise.resolve(userData));
      await login.instance().handleSubmit(mockEvent);

      expect(fetchUserData).toHaveBeenCalledTimes(1);
    });

    it('call alert if user is not found', async () => {
      login.setState({ email: 'asdf@aol.com' });
      const mockEvent = { preventDefault: jest.fn() };
      window.alert = jest.fn();
      const userData = {
        data: [
          {
            id: 1,
            name: 'Taylor',
            password: 'password',
            email: 'tman2272@aol.com'
          }
        ]
      };
      fetchUserData.mockImplementation(() => Promise.resolve(userData));
      await login.instance().handleSubmit(mockEvent);

      expect(window.alert).toHaveBeenCalledTimes(1);
      expect(window.alert).toHaveBeenCalledWith('No such user');
    });

    it("calls alert user is found, but password doesn't match", async () => {
      login.setState({ email: 'tman2272@aol.com', password: 'asdf' });
      const mockEvent = { preventDefault: jest.fn() };
      window.alert = jest.fn();
      const userData = {
        data: [
          {
            id: 1,
            name: 'Taylor',
            password: 'password',
            email: 'tman2272@aol.com'
          }
        ]
      };
      fetchUserData.mockImplementation(() => Promise.resolve(userData));
      await login.instance().handleSubmit(mockEvent);

      expect(window.alert).toHaveBeenCalledTimes(1);
      expect(window.alert).toHaveBeenCalledWith('Incorrect Password');
    });

    it('calls fetchFavoriteData is user is validated', async () => {
      login.setState({ email: 'tman2272@aol.com', password: 'password' });
      const mockEvent = { preventDefault: jest.fn() };
      window.alert = jest.fn();
      const userData = {
        data: [
          {
            id: 1,
            name: 'Taylor',
            password: 'password',
            email: 'tman2272@aol.com'
          }
        ]
      };
      fetchUserData.mockImplementation(() => Promise.resolve(userData));
      await login.instance().handleSubmit(mockEvent);

      expect(fetchFavoriteData).toHaveBeenCalledTimes(1);
    });

    it('calls toggleLogin is user is validated', async () => {
      login.setState({ email: 'tman2272@aol.com', password: 'password' });
      const mockEvent = { preventDefault: jest.fn() };
      window.alert = jest.fn();
      const userData = {
        data: [
          {
            id: 1,
            name: 'Taylor',
            password: 'password',
            email: 'tman2272@aol.com'
          }
        ]
      };
      fetchUserData.mockImplementation(() => Promise.resolve(userData));
      await login.instance().handleSubmit(mockEvent);

      expect(mockProps.toggleLogin).toHaveBeenCalledTimes(1);
    });

    it('calls storeUserData is user is validated', async () => {
      login.setState({ email: 'tman2272@aol.com', password: 'password' });
      const mockEvent = { preventDefault: jest.fn() };
      window.alert = jest.fn();
      const userData = {
        data: [
          {
            id: 1,
            name: 'Taylor',
            password: 'password',
            email: 'tman2272@aol.com'
          }
        ]
      };
      const foundUser = {
        id: 1,
        name: 'Taylor',
        password: 'password',
        email: 'tman2272@aol.com'
      };
      fetchUserData.mockImplementation(() => Promise.resolve(userData));
      await login.instance().handleSubmit(mockEvent);

      expect(mockProps.storeUserData).toHaveBeenCalledTimes(1);
      expect(mockProps.storeUserData).toHaveBeenCalledWith(foundUser);
    });

    it('calls loadFavorites is user is validated', async () => {
      login.setState({ email: 'tman2272@aol.com', password: 'password' });
      const mockEvent = { preventDefault: jest.fn() };
      window.alert = jest.fn();
      const userData = {
        data: [
          {
            id: 1,
            name: 'Taylor',
            password: 'password',
            email: 'tman2272@aol.com'
          }
        ]
      };
      const favoritesArray = [];
      fetchUserData.mockImplementation(() => Promise.resolve(userData));
      fetchFavoriteData.mockImplementation(() =>
        Promise.resolve(favoritesArray)
      );
      await login.instance().handleSubmit(mockEvent);

      expect(mockProps.loadFavorites).toHaveBeenCalledTimes(1);
      expect(mockProps.loadFavorites).toHaveBeenCalledWith(favoritesArray);
    });

    it('calls loadFavorites is user is validated', async () => {
      login.setState({ email: 'tman2272@aol.com', password: 'password' });
      const mockEvent = { preventDefault: jest.fn() };
      window.alert = jest.fn();
      const userData = {
        data: [
          {
            id: 1,
            name: 'Taylor',
            password: 'password',
            email: 'tman2272@aol.com'
          }
        ]
      };
      const favoritesArray = [];
      fetchUserData.mockImplementation(() => Promise.resolve(userData));
      fetchFavoriteData.mockImplementation(() =>
        Promise.resolve(favoritesArray)
      );
      await login.instance().handleSubmit(mockEvent);

      expect(mockProps.history.push).toHaveBeenCalledTimes(1);
      expect(mockProps.history.push).toHaveBeenCalledWith('/');
    });
  });

  describe('mapStateToProps', () => {
    it('should map state properties to props', () => {
      const mockState = {
        user: {
          loggedIn: false,
          favorites: []
        }
      };
      const expected = {
        loggedIn: false,
        favorites: []
      };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('calls dispatch on toggleLogin', () => {
      const mockDispatch = jest.fn();
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = toggleLogin();
      mappedProps.toggleLogin();

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('calls dispatch with the correct params on storeUserData', () => {
      const mockDispatch = jest.fn();
      const userData = {};
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = storeUserData(userData);
      mappedProps.storeUserData(userData);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('calls dispatch with the correct params on loadFavorites', () => {
      const mockDispatch = jest.fn();
      const mockFavoritesArray = [];
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = loadFavorites(mockFavoritesArray);
      mappedProps.loadFavorites(mockFavoritesArray);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });
  });
});
