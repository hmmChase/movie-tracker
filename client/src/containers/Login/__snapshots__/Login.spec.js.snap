// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    favorites={Array []}
    history={
        Object {
            "push": [Function],
          }
    }
    loadFavorites={[Function]}
    loggedIn={false}
    storeUserData={[Function]}
    toggleLogin={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <input
                className="inputField"
                name="email"
                onChange={[Function]}
                placeholder="Enter Email"
                type="text"
                value=""
        />
        <input
                className="inputField"
                name="password"
                onChange={[Function]}
                placeholder="Enter Password"
                type="password"
                value=""
        />
        <button
                className="loginButton"
        />
</form>,
      "className": "Login",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="inputField"
            name="email"
            onChange={[Function]}
            placeholder="Enter Email"
            type="text"
            value=""
/>,
          <input
            className="inputField"
            name="password"
            onChange={[Function]}
            placeholder="Enter Password"
            type="password"
            value=""
/>,
          <button
            className="loginButton"
/>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "inputField",
            "name": "email",
            "onChange": [Function],
            "placeholder": "Enter Email",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "inputField",
            "name": "password",
            "onChange": [Function],
            "placeholder": "Enter Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "loginButton",
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <input
                    className="inputField"
                    name="email"
                    onChange={[Function]}
                    placeholder="Enter Email"
                    type="text"
                    value=""
          />
          <input
                    className="inputField"
                    name="password"
                    onChange={[Function]}
                    placeholder="Enter Password"
                    type="password"
                    value=""
          />
          <button
                    className="loginButton"
          />
</form>,
        "className": "Login",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="inputField"
              name="email"
              onChange={[Function]}
              placeholder="Enter Email"
              type="text"
              value=""
/>,
            <input
              className="inputField"
              name="password"
              onChange={[Function]}
              placeholder="Enter Password"
              type="password"
              value=""
/>,
            <button
              className="loginButton"
/>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "inputField",
              "name": "email",
              "onChange": [Function],
              "placeholder": "Enter Email",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "inputField",
              "name": "password",
              "onChange": [Function],
              "placeholder": "Enter Password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "loginButton",
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
